File: 2015/day11/day11.py
def checkTriplets(s):
    i = 0
    while i < len(s) - 2:
        if ord(s[i+1]) == ord(s[i]) + 1 and ord(s[i+2]) == ord(s[i]) + 2:
            return True
        else:
            i += 1
    return False

def checkDoubles(s):
    count = 0
    i = 0
    while i < len(s) - 1:
        forbidden_chars = ['i', 'o', 'l']
        if s[i] in forbidden_chars:
            return False
        elif s[i] == s[i+1] and (i == len(s) - 2 or s[i+1] != s[i+2]):
            count += 1
        i += 1
    return count >= 2

def generatePassword(s):
    i = True
    while i:
        if checkDoubles(s) and checkTriplets(s):
            i = False
        else:
            index = len(s) - 1
            while index >= 0 and s[index] == 'z':
                index -= 1
            if index >= 0:
                s = s[:index] + chr(ord(s[index]) + 1) + 'a' * (len(s) - index - 1)
            else:
                s = 'a' + s
    return s

str = "hepxcrrq"
print("Part one: " + generatePassword(str))
str1 = "hepxxzaa"
print("Part two: " + generatePassword(str1))

File: 2015/day11/day11.cpp
#include<iostream>
#include<stdlib.h>
using namespace std;

bool checkTriplets(string str){
    int i = 0;
    while(str[i] != '\0' && str[i+1] != '\0' && str[i+2] != '\0'){
        if(str[i+1] == str[i] + 1 && str[i+2] == str[i] + 2){
            return true;
        } else {
            i++;
        }
    }
    return false;
}

bool checkDoubles(string str){
    int count = 0;
    int i = 0;
    while(str[i] != '\0' && str[i+1] != '\0'){
        if(str[i] == 'i' || str[i] == 'o' || str[i] == 'l'){
            return false;
        } else if((str[i] == str[i+1]) && str[i+1] != str[i+2]){
            count++;
        }
        i++;
    }
    return (count >= 2) ? true : false;
}

string generatePassword(string str) {
    int size = str.size();
    bool i = true;
    while (i) {
        if (checkDoubles(str) && checkTriplets(str)) {
            i = false;
        } else {
            int index = size - 1;
            while (index >= 0 && str[index] == 'z') {
                str[index] = 'a';
                index--;
            }
            if (index >= 0)
                str[index]++;
            else
                str.insert(0, "a");
        }
    }
    return str;
}

int main(){
    string str = "hepxcrrq";
    string passwordPartOne = generatePassword(str);
    cout<<"Part One: " <<passwordPartOne<<endl;
    string str1 = "hepxxzaa";
    string passwordPartTwo = generatePassword(str1);
    cout<<"Part Two: " <<passwordPartTwo<<endl;
    return 0;
}
